bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,467 @@
+        """
+        if (symbol := self._results.get(doc_item)) is not None:
+            return symbol
+
+        if (symbols_to_queue := self._page_symbols.get(doc_item.url)) is not None:
+            async with client_session.get(doc_item.url) as response:
+                soup = BeautifulSoup(await response.text(encoding="utf8"), "lxml")
+
+            self._queue.extend(QueueItem(symbol, soup) for symbol in symbols_to_queue)
+            del self._page_symbols[doc_item.url]
+            log.debug(f"Added symbols from {doc_item.url} to parse queue.")
+
+            if self._parse_task is None:
+                self._parse_task = asyncio.create_task(self._parse_queue())
+
+        self._move_to_front(doc_item)
+        self._item_events[doc_item] = item_event = asyncio.Event()
+        await item_event.wait()
+        return self._results[doc_item]
+
+    async def _parse_queue(self) -> None:
+        """
+        Parse all item from the queue, setting associated events for symbols if present.
+
+        The coroutine will run as long as the queue is not empty, resetting `self._parse_task` to None when finished.
+        """
+        log.trace("Starting queue parsing.")
+        while self._queue:
+            item, soup = self._queue.pop()
+            try:
+                markdown = get_symbol_markdown(soup, item)
+                await doc_cache.set_if_exists(item, markdown)
+                self._results[item] = markdown
+            except Exception:
+                log.exception(f"Unexpected error when handling {item}")
+            else:
+                if (event := self._item_events.get(item)) is not None:
+                    event.set()
+            await asyncio.sleep(0.1)
+
+        self._parse_task = None
+        log.trace("Finished parsing queue.")
+
+    def _move_to_front(self, item: Union[QueueItem, DocItem]) -> None:
+        """Move `item` to the front of the parse queue."""
+        # The parse queue stores soups along with the doc symbols in QueueItem objects,
+        # in case we're moving a DocItem we have to get the associated QueueItem first and then move it.
+        item_index = self._queue.index(item)
+        queue_item = self._queue[item_index]
+
+        del self._queue[item_index]
+        self._queue.append(queue_item)
+
+    def add_item(self, doc_item: DocItem) -> None:
+        """Add a DocItem to `_page_symbols`."""
+        self._page_symbols[doc_item.url].append(doc_item)
+
+    async def clear(self) -> None:
+        """
+        Clear all internal symbol data.
+
+        All currently requested items are waited to be parsed before clearing.
+        """
+        for event in self._item_events.values():
+            await event.wait()
+        if self._parse_task is not None:
+            self._parse_task.cancel()
+            self._parse_task = None
+        self._queue.clear()
+        self._results.clear()
+        self._page_symbols.clear()
+        self._item_events.clear()
```

Comment:
There may be a race condition here. While clearing I believe you need to lock to ensure that no new items can be queued.

Reply:
The code is being ran synchronously so nothing yields control to the event loop.

But I had a concern that I forgot about with the cancel here as I'm not sure how exactly it works. The docs mention that the exception will be raised at the next cycle of the event loop. Does that mean if this was getting executed while _parse_queue is suspended It'll be cancelled immediately when it gets control or on the next await inside of it?

Reply:
> The code is being ran synchronously so nothing yields control to the event loop.

Each event is awaited. That provides many opportunities to yield control. It may also be an issue that the events dictionary could get updated during iteration.

> Does that mean if this was getting executed while _parse_queue is suspended It'll be cancelled immediately when it gets control or on the next await inside of it?

Not sure. Probably the former.

Reply:
Oh right I was focused on the clears themselves when thinking about it and skipped over the waits at the start. Will lock the commands when a refresh is taking place

Reply:
Implemented the locks in ac6d25cf, now every coro that can trigger a refresh is behind a lock and the get command will wait for refreshes to finish before proceeding to schedule new symbols